File: last/last v2/03-intro-java/Colabjava/Example11/Example11.java
// =================================================================
//
// File: Example11.java
// Author(s):
// Description: This file implements the code that transforms a
//				grayscale image. The time this implementation takes will
//				be used as the basis to calculate the improvement obtained
// 				with parallel technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

/*----------------------------------------------------------------

*

* Multiprocesadores: Java

* Fecha: 10-Sep-2021

* Autor: A01209400 Royer Donnet Arenas Camacho
		 A01654856 Hugo David Franco Avila

*

*--------------------------------------------------------------*/

import java.awt.image.BufferedImage;
import java.io.File;
import javax.imageio.ImageIO;

public class Example11 {
	private int src[], dest[], width, height;

	public Example11(int src[], int dest[], int width, int height) {
		this.src = src;
		this.dest = dest;
		this.width = width;
		this.height = height;
	}

	// palce your code here

	private int scaleToGray(int color){
		int blue = color & 0xff;
		int green = (color & 0xff00) >> 8;
		int red = (color & 0xff0000) >> 16;
		int gray = (blue + green + red)/3;
		int alpha = (color & 0xff000000) >>> 24;
		int newPixel = (alpha & 0x000000ff) << 24 | (gray & 0x000000ff) << 16 | (gray & 0x000000ff) << 8 | (gray & 0x000000ff);
		return newPixel;
	}

	public void doTask() {
		for(int i=0;i<src.length;i++){
			dest[i]=scaleToGray(src[i]);
		}
	}

	public static void main(String args[]) throws Exception {
		long startTime, stopTime;
		double ms;

		if (args.length != 1) {
			System.out.println("usage: java Example11 image_file");
			System.exit(-1);
		}

		final String fileName = args[0];
		File srcFile = new File(fileName);
        final BufferedImage source = ImageIO.read(srcFile);

		int w = source.getWidth();
		int h = source.getHeight();
		int src[] = source.getRGB(0, 0, w, h, null, 0, w);
		int dest[] = new int[src.length];

		System.out.printf("Starting...\n");
		ms = 0;
		Example11 e = new Example11(src, dest, w, h);
		for (int i = 0; i < Utils.N; i++) {
			startTime = System.currentTimeMillis();

			e.doTask();

			stopTime = System.currentTimeMillis();

			ms += (stopTime - startTime);
		}
		System.out.printf("avg time = %.5f\n", (ms / Utils.N));
		final BufferedImage destination = new BufferedImage(w, h, BufferedImage.TYPE_INT_ARGB);
		destination.setRGB(0, 0, w, h, dest, 0, w);

		javax.swing.SwingUtilities.invokeLater(new Runnable() {
            public void run() {
               ImageFrame.showImage("Original - " + fileName, source);
            }
        });

		javax.swing.SwingUtilities.invokeLater(new Runnable() {
            public void run() {
               ImageFrame.showImage("Gray - " + fileName, destination);
            }
        });
	}
}

File: act col/examples/06-openmp/colab6/example11.cpp
// =================================================================
//
// File: example11.cpp
// Author(s):
// Description: This file implements the code that transforms a
//				grayscale image. Using OpenCV and OpenMP, to compile:
//				g++ example11.cpp `pkg-config --cflags --libs opencv4` -fopenmp
//
//				The time this implementation takes will be used as the
//				basis to calculate the improvement obtained with
//				parallel technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

/*----------------------------------------------------------------

*

* Multiprocesadores: OpenMP

* Fecha: 12-oct-2021

* Autor: A01209400 - Royer Donnet Arenas Camacho
		 A01654856 - Hugo David Franco Ávila 

*

*--------------------------------------------------------------*/

#include <stdio.h>
#include <stdlib.h>
#include <opencv2/opencv.hpp>
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/highgui/highgui_c.h>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/imgcodecs/imgcodecs.hpp>
#include "utils.h"

// implement your code
class GrayScale
{
private:
	cv::Mat &src, &dest;

	void scaleToGray(int ren, int col) const {
		float r, g, b;

		r = 0;
		g = 0;
		b = 0;
		r += (float)src.at<cv::Vec3b>(ren, col)[RED];
		g += (float)src.at<cv::Vec3b>(ren, col)[GREEN];
		b += (float)src.at<cv::Vec3b>(ren, col)[BLUE];

		float gray = (r + g + b)/3.0;

		dest.at<cv::Vec3b>(ren, col)[RED] = (unsigned char)(gray);
		dest.at<cv::Vec3b>(ren, col)[GREEN] = (unsigned char)(gray);
		dest.at<cv::Vec3b>(ren, col)[BLUE] = (unsigned char)(gray);
	}

	public:
		GrayScale(cv::Mat &s, cv::Mat &d) : src(s), dest(d) {}

		void doTask() {
			int i,j;
			#pragma omp parallel for private(j)
			for (i = 0; i < src.rows; i++) {
				for (j = 0; j < src.cols; j++) {
					scaleToGray(i, j);
				}
			}
		}
};

int main(int argc, char *argv[])
{
	int i;
	double acum;

	if (argc != 2)
	{
		printf("usage: %s source_file\n", argv[0]);
		return -1;
	}

	cv::Mat src = cv::imread(argv[1], cv::IMREAD_COLOR);
	cv::Mat dest = cv::Mat(src.rows, src.cols, CV_8UC3);
	if (!src.data)
	{
		printf("Could not load image file: %s\n", argv[1]);
		return -1;
	}

	acum = 0;
	for (i = 0; i < N; i++)
	{
		start_timer();

		GrayScale obj(src, dest);
		obj.doTask();

		acum += stop_timer();
	}

	printf("avg time = %.5lf ms\n", (acum / N));

	/*
	cv::namedWindow("Original", cv::WINDOW_AUTOSIZE);
	cv::imshow("Original", src);

	cv::namedWindow("Gray scale", cv::WINDOW_AUTOSIZE);
	cv::imshow("Gray scale", dest);

	cv::waitKey(0);
	*/

	cv::imwrite("gray_scale.png", dest);

	return 0;
}
